src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,429 @@
+
+@[simp] lemma zero_def : (0 : language α) = ∅ := rfl
+@[simp] lemma one_def : (1 : language α) = {[]} := rfl
+
+@[simp] lemma add_def (l m : language α) : l + m = l ∪ m := rfl
+@[simp] lemma mul_def (l m : language α) : l * m = (l.prod m).image (λ p, p.1 ++ p.2) := rfl
+
+/-- The star of a language `L` is the set of all strings which can be written by concatenating
+  strings from `L`. -/
+def star (l : language α) : language α :=
+{ x | ∃ S : list (list α), x = S.join ∧ ∀ y ∈ S, ¬(list.empty y) ∧ y ∈ l}
+
+@[simp] lemma star_def (l : language α) :
+  l.star = { x | ∃ S : list (list α), x = S.join ∧ ∀ y ∈ S, ¬(list.empty y) ∧ y ∈ l} := rfl
+
+lemma mul_assoc (l m n : language α) : (l * m) * n = l * (m * n) :=
+begin
+  ext x,
+  simp only [mul_def, exists_and_distrib_left, set.mem_image2, set.image_prod],
+  split,
+  { rintro ⟨ a, ⟨ b, hb, c, hc, ha⟩, d, hd, hx ⟩,
+    refine ⟨ b, hb, c ++ d, ⟨ c, hc, d, hd, rfl ⟩, _ ⟩,
+    rw [←list.append_assoc, ha, hx] },
+  { rintro ⟨ a, ha, b, ⟨ c, hc, d, hd, hb ⟩, hx ⟩,
+    refine ⟨ a ++ c, ⟨ a, ha, c, hc, rfl ⟩, d, hd, _ ⟩,
+    rw [list.append_assoc, hb, hx] }
+end
+
+lemma one_mul (l : language α) : 1 * l = l :=
+begin
+  ext x,
+  simp only [mul_def, set.mem_image, prod.mk.inj_iff, one_def, set.singleton_prod, prod.exists],
+  split,
+  { rintro ⟨ a, b, ⟨ c, hc, hnil, hx ⟩, _ ⟩,
+    finish },
+  { intro hx,
+    exact ⟨ [], x, ⟨ x, hx, rfl, rfl ⟩, rfl ⟩ }
+end
+
+lemma mul_one (l : language α) : l * 1 = l :=
+begin
+  ext x,
+  simp only [mul_def, set.mem_image, prod.mk.inj_iff, one_def, set.prod_singleton, prod.exists],
+  split,
+  { rintro ⟨ a, b, ⟨ c, hc, ha, hnil ⟩, hx ⟩,
+    rw [←hnil, ←ha, list.append_nil] at hx,
+    rwa ←hx },
+  { intro hx,
+    exact ⟨ x, [], ⟨ x, hx, rfl, rfl ⟩, x.append_nil ⟩ }
+end
+
+lemma left_distrib (l m n : language α) : l * (m + n) = (l * m) + (l * n) :=
+begin
+  ext x,
+  simp only [mul_def, set.mem_image, add_def, set.mem_prod, exists_and_distrib_left, set.mem_image2,
+    set.image_prod, set.mem_union_eq, set.prod_union, prod.exists],
+  split,
+  { rintro ⟨ y, z, (⟨ hy, hz ⟩ | ⟨ hy, hz ⟩), hx ⟩,
+    { left,
+      exact ⟨ y, hy, z, hz, hx ⟩ },
+    { right,
+      exact ⟨ y, hy, z, hz, hx ⟩ } },
+  { rintro (⟨ y, hy, z, hz, hx ⟩ | ⟨ y, hy, z, hz, hx ⟩);
+    refine ⟨ y, z, _, hx ⟩,
+    { left,
+      exact ⟨ hy, hz ⟩ },
+    { right,
+      exact ⟨ hy, hz ⟩ } }
+end
+
+lemma right_distrib (l m n : language α) : (l + m) * n = (l * n) + (m * n) :=
+begin
+  ext x,
+  simp only [mul_def, set.mem_image, add_def, set.mem_prod, exists_and_distrib_left, set.mem_image2,
+    set.image_prod, set.mem_union_eq, set.prod_union, prod.exists],
+  split,
+  { rintro ⟨ y, (hy | hy), z, hz, hx ⟩,
+    { left,
+      exact ⟨ y, hy, z, hz, hx ⟩ },
+    { right,
+      exact ⟨ y, hy, z, hz, hx ⟩ } },
+  { rintro (⟨ y, hy, z, hz, hx ⟩ | ⟨ y, hy, z, hz, hx ⟩);
+    refine ⟨ y, _, z, hz, hx ⟩,
+    { left,
+      exact hy },
+    { right,
+      exact hy } }
+end
```

Comment:
It probably makes sense to mark these lemmas `private`, since no one should be using them directly, and instead going through `semiring`